File: 2020/julia/src/Day4.jl
module Day4
include("Aoc.jl")
using .Aoc

required_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
eye_colors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

data = Aoc.input_string(4)

passports = map(split(data, r"\n\n")) do pd
    passport = Dict{String, String}()
    for field in split(strip(pd), r"[\n\s]")
        (key, val) = split(String(field), ":")
        passport[key] = val
    end
    passport
end

valid_passports = filter(passport -> findfirst(req -> !haskey(passport, req), required_fields) == nothing, passports)

year(val, min, max) = min <= parse(Int64, val) <= max

function height(input)
    m = match(r"^(?<len>[0-9]{2,3})(?<unit>in|cm)$", input)
    m != nothing && (
        (m[:unit] == "cm" && 150 <= parse(Int32, m[:len]) <= 193) 
        || (m[:unit] == "in" && 59 <= parse(Int32, m[:len]) <= 76)
    )
end

part1() = size(valid_passports)[1]
part2() = valid_passports|>
    list -> filter(passport -> year(passport["byr"], 1920, 2002), list)|>
    list -> filter(passport -> year(passport["iyr"], 2010, 2020), list)|>
    list -> filter(passport -> year(passport["eyr"], 2020, 2030), list)|>
    list -> filter(passport -> height(passport["hgt"]), list)|>
    list -> filter(passport -> occursin(r"^\#[0-9a-f]{6}$", passport["hcl"]), list)|>
    list -> filter(passport -> passport["ecl"] ∈ eye_colors, list)|>
    list -> filter(passport -> occursin(r"^[0-9]{9}$", passport["pid"]), list)|>
    size|>
    first
end

File: 2020/typescript/src/days/day4.ts
import {Day} from "../aoc";

export class Day4 extends Day {
    private passports: Map<string, string>[];

    private readonly required = new Map<string, (i: string) => boolean>([
        ['byr', (i) => Day4.year(i, 1920, 2002)],
        ['iyr',  (i) => Day4.year(i, 2010, 2020)],
        ['eyr',  (i) => Day4.year(i, 2020, 2030)],
        ['hgt', (i) => Day4.height(i)],
        ['hcl', (i) => i.match(/^#[a-f0-9]{6}$/) !== null],
        ['ecl', (i) => i.match(/^(amb|blu|brn|gry|grn|hzl|oth)$/) !== null],
        ['pid', (i) => i.match(/^[0-9]{9}$/) !== null],
    ]);

    day = (): number => 4;

    part1 = (): string => this.passports.filter(p => this.hasRequiredFields(p)).length.toString();

    part2 = (): string => this.passports.filter(p => this.hasValidFields(p)).length.toString();

    setup = () =>
        this.passports = this.input
            .join("\n")
            .split("\n\n")
            .map(s => s.split(/\s+/).map(p => p.split(':', 2)))
            .map(document => {
                const map = new Map<string, string>();
                document.forEach(field => map.set(field[0], field[1]));
                return map;
            });

    private hasRequiredFields(passport: Map<string, string>): boolean {
        for (const field of this.required.keys()) {
            if (!passport.has(field)) return false;
        }
        return true;
    }

    private hasValidFields(passport: Map<string, string>): boolean {
        for(const [field, validator] of this.required.entries()) {
            if(!passport.has(field) || !validator(passport.get(field))) return false;
        }
        return true;
    }

    private static year(i: string, min: number, max: number): boolean {
        const n = parseInt(i);
        return !isNaN(n) && n >= min && n <= max;
    }

    private static height(i: string): boolean {
        const m = i.match(/^(?<len>[0-9]{2,3})(?<unit>in|cm)$/);
        if(m === null) return false;
        const [unit, len] = [m.groups.unit, parseInt(m.groups.len)];
        return (unit === 'cm' && len >= 150 && len <= 193) || (unit === 'in' && len >= 59 && len <= 76);
    }
}